datafusion/expr/src/udf.rs

```diff
@@ -249,6 +261,22 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// the arguments
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType>;
 
+    /// What [`DataType`] will be returned by this function, given the types of
+    /// the expr arguments
+    fn return_type_from_exprs(
```

Comment:
Since we need to use the `ExprSchema` here, but there's issues going from `<S: ExprSchema>` to `&dyn ExprSchema` inside the trait here, what if we moved up the default implementation out of the trait and into `ScalarUDF`? 

we could change the trait impl to something like this

```rs
pub trait ScalarUDFImpl: Debug + Send + Sync {
    /// What [`DataType`] will be returned by this function, given the types of
    /// the expr arguments
    fn return_type_from_exprs(
        &self,
        arg_exprs: &[Expr],
        schema: &dyn ExprSchema,
    ) -> Option<Result<DataType>> {
        // The default implementation returns None
        // so that people don't have to implement `return_type_from_exprs` if they dont want to
        None
    }
}
```
then change the `ScalarUDF` impl to this
```rs
impl ScalarUDF
    /// The datatype this function returns given the input argument input types.
    /// This function is used when the input arguments are [`Expr`]s.
    /// See [`ScalarUDFImpl::return_type_from_exprs`] for more details.
    pub fn return_type_from_exprs<S: ExprSchema>(
        &self,
        args: &[Expr],
        schema: &S,
    ) -> Result<DataType> {
        // If the implementation provides a return_type_from_exprs, use it
        if let Some(return_type) = self.inner.return_type_from_exprs(args, schema) {
            return_type
        // Otherwise, use the return_type function
        } else {
            let arg_types = args
                .iter()
                .map(|arg| arg.get_type(schema))
                .collect::<Result<Vec<_>>>()?;
            self.return_type(&arg_types)
        }
    }
}
```

this way all constraints are met, we can update the `get_type` function to use the `return_type_from_exprs`

```rs
ScalarFunctionDefinition::UDF(fun) => {
    Ok(fun.return_type_from_exprs(&args, schema)?)
}
```

and it still makes `return_type_from_exprs` an opt-in method.

Reply:
This works well on my side. Thanks!

Another question for me is, how can user implement `return_type_from_exprs` when using `schema`. For example, `arg_exprs.get(take_idx).unwrap().get_type(schema)` will lead an error

> the size for values of type `dyn ExprSchema` cannot be known at compilation time
> the trait `Sized` is not implemented for `dyn ExprSchema`

Reply:
Hmm yeah @yyy1000 you still run into the same error then.



I'm wondering if it'd be easiest to just change the type signature on `ExprSchemable` to be generic over the trait instead of the functions. 

```rs
pub trait ExprSchemable<S: ExprSchema> {
    /// given a schema, return the type of the expr
    fn get_type(&self, schema: &S) -> Result<DataType>;

    /// given a schema, return the nullability of the expr
    fn nullable(&self, input_schema: &S) -> Result<bool>;

    /// given a schema, return the expr's optional metadata
    fn metadata(&self, schema: &S) -> Result<HashMap<String, String>>;

    /// convert to a field with respect to a schema
    fn to_field(&self, input_schema: &DFSchema) -> Result<DFField>;

    /// cast to a type with respect to a schema
    fn cast_to(self, cast_to_type: &DataType, schema: &S) -> Result<Expr>;
}

impl ExprSchemable<DFSchema> for Expr {
//... 
}
```


then the trait can just go back to the original implementation you had using `&DFSchema`

```rs
    fn return_type_from_exprs(
        &self,
        arg_exprs: &[Expr],
        schema: &DFSchema,
    ) -> Result<DataType> {
        let arg_types = arg_exprs
            .iter()
            .map(|e| e.get_type(schema))
            .collect::<Result<Vec<_>>>()?;
        self.return_type(&arg_types)
    }
```

I tried this locally and was able to get things to compile locally, and was able to implement a udf using the trait. 

It does make it a little less flexible as it's expecting a `DFSchema`, but i _think_ thats ok? 

I think the only other approach would be to make `ScalarUDFImpl` dynamic over `<S: ExprSchema>`, but I feel like that's much less ideal than just using a concrete type.

Reply:
Thanks for your help! @universalmind303 
It looks good and I think I can try it to see. :)

Reply:
Update is I changed the signature to take `&dyn ExprSchema` which seems to have worked just fine